lib/1.4/dml-builtins.dml

```diff
@@ -2825,136 +2826,89 @@ template register is (_conf_attribute, get, set, shown_desc,
-        foreach f in (this.fields) {
+        local uint64 lsbs = 0;
+
+        // counting sort
+        foreach f in (_get_fields) {
             local int lsb = f.lsb;
-            // sort 'fields' by lsb
-            local int i;
-            for (i = n; i > 0; --i) {
-                local int next_lsb = fields[i - 1].lsb;
-                if (next_lsb > lsb) {
-                    fields[i] = fields[i - 1];
-                } else {
-                    break;
-                }
-            }
-            fields[i] = f;
+            lsbs[lsb] = 1;
             unmapped[lsb + f.bitsize - 1 : lsb] = 0;
-            ++n;
+            n += 1;
         }
-        assert n <= 64;
-        return (n, unmapped);
-    }
 
-    // fills in the passed array with all fields in this register, and returns
-    // the number of elements and a bitmask showing bits not covered by fields.
-    shared method _get_get_fields(_get_field *fields) -> (int, uint64) /* n, unmapped */ {
-        local uint64 unmapped;
-        unmapped[this.bitsize-1:0] = -1;
-        local int n = 0;
-        // TODO: it is inefficient to re-generate the field list each call
-        foreach f in (this._get_fields) {
-            local int lsb = f.lsb;
-            // sort 'fields' by lsb
-            local int i;
-            for (i = n; i > 0; --i) {
-                local int next_lsb = fields[i - 1].lsb;
-                if (next_lsb > lsb) {
-                    fields[i] = fields[i - 1];
-                } else {
-                    break;
-                }
-            }
-            fields[i] = f;
-            unmapped[lsb + f.bitsize - 1 : lsb] = 0;
-            ++n;
+        foreach f in (_get_fields) {
+            fields[bit_count64(lsbs[f.lsb : 0]) - 1] = f;
         }
-        assert n <= 64;
         return (n, unmapped);
     }
 
     // fills in the passed array with all fields in this register, and returns
     // the number of elements and a bitmask showing bits not covered by fields.
     shared method _get_set_fields(_set_field *fields) -> (int, uint64) /* n, unmapped */ {
         local uint64 unmapped;
-        unmapped[this.bitsize-1:0] = -1;
+        unmapped[this.bitsize - 1 : 0 ] = -1;
         local int n = 0;
-        // TODO: it is inefficient to re-generate the field list each call
-        foreach f in (this._set_fields) {
+        local uint64 lsbs = 0;
+
+        // counting sort
+        foreach f in (_set_fields) {
             local int lsb = f.lsb;
-            // sort 'fields' by lsb
-            local int i;
-            for (i = n; i > 0; --i) {
-                local int next_lsb = fields[i - 1].lsb;
-                if (next_lsb > lsb) {
-                    fields[i] = fields[i - 1];
-                } else {
-                    break;
-                }
-            }
-            fields[i] = f;
+            lsbs[lsb] = 1;
             unmapped[lsb + f.bitsize - 1 : lsb] = 0;
-            ++n;
+            n += 1;
```

Comment:
Instead of accumulating `n` in a loop, use `_set_fields.len`, instead OR `bit_count64(lsbs)`. The latter is only a single cycle, rather than some cycles per number of fields. Again, since the entire point of this PR is to optimize, I'm going to speak in favor of it unless you can show through benches that it makes no difference.

Reply:
Nope! Accumulating `n` is faster! Slightly faster than returning `bit_count64(lsbs)`, and much faster than `_set_fields.len`. My guess is that it gets ILP:d away.

Reply:
Surprisingly `n++` is about as fast as `n += 1` even though the generation of `n += 1` takes the address of `n` and `n++` doesn't.

Reply:
Oo! If you get replace `n += 1` with `++n` you'll cut down on the diff since `++n` was the original code!

Reply:
Yeah remove this diff